### ISSUE TITLE
perfetto collection issue with configuration.

### ISSUE THREAD

**Issue by forrestxu123** [google/perfetto#864]
when i use  command below:

adb push config.pbtx /data/local/tmp/config.pbtx
adb shell 'cat /data/local/tmp/config.pbtx | perfetto --txt -c - -o /data/misc/perfetto-traces/trace.perfetto-trace'

Some times  i will got 0 size of file trace.perfetto-trace if duration_ms: is set to 50000 (seems 40000) works fine. what is the reason? see the config.pbtx below:

buffers: {
    size_kb: 1522240
    fill_policy: DISCARD
}
buffers: {
    size_kb: 4096
    fill_policy: DISCARD
}
data_sources: {
    config {
        name: "android.log"
        android_log_config {
        }
    }
}

data_sources: {
    config {
        name: "android.packages_list"
        target_buffer: 1
    }
}
data_sources: {
    config {
        name: "linux.process_stats"
        target_buffer: 1
        process_stats_config {
            scan_all_processes_on_start: true
        }
    }
}
data_sources: {
    config {
        name: "linux.perf"
        perf_event_config {
            timebase {
                frequency: 4000
                timestamp_clock: PERF_CLOCK_BOOTTIME
            }
            callstack_sampling {
            }
        }
    }
}
data_sources: {
    config {
        name: "linux.sys_stats"
        sys_stats_config {
            stat_period_ms: 1000
            stat_counters: STAT_CPU_TIMES
            stat_counters: STAT_FORK_COUNT
            cpufreq_period_ms: 1000
        }
    }
}

data_sources: {
    config {
        name: "linux.ftrace"
        ftrace_config {
            ftrace_events: "sched/sched_switch"
            ftrace_events: "power/suspend_resume"
            ftrace_events: "sched/sched_wakeup"
            ftrace_events: "sched/sched_wakeup_new"
            ftrace_events: "sched/sched_waking"
            ftrace_events: "power/cpu_frequency"
            ftrace_events: "power/cpu_idle"
            ftrace_events: "raw_syscalls/sys_enter"
            ftrace_events: "raw_syscalls/sys_exit"
            ftrace_events: "sched/sched_process_exit"
            ftrace_events: "sched/sched_process_free"
            ftrace_events: "task/task_newtask"
            ftrace_events: "task/task_rename"
        }
    }
}
duration_ms: 40000

**Comment by LalitMaganti** [comment]
`size_kb: 1522240` is allocating a buffer of 1.5GB. It's very likely that you are killing the system by just writing into too much memory: Perfetto by default keeps data in RAM unless you use [long trace mode](https://perfetto.dev/docs/concepts/config#long-traces) to stream the data to disk.

In summary, you probably need to add:
```
write_into_file: true
file_write_period_ms: 5000
```

It's very very strongly recommended that you use RING_BUFFER configs for write_into_file traces: using DISCARD buffer is not well defined.

**Comment by forrestxu123** [comment]
using  DISCARD can not see callstack sample

**Comment by LalitMaganti** [comment]
I responded to this at https://github.com/google/perfetto/issues/863

**Comment by forrestxu123** [comment]
WHERE TO ADD:write_into_file: true
file_write_period_ms: 5000
THANKS

    On Tuesday, August 20, 2024 at 03:16:51 PM PDT, Lalit Maganti ***@***.***> wrote:  
 
 


size_kb: 1522240 is allocating a buffer of 1.5GB. It's very likely that you are killing the system by just writing into too much memory: Perfetto by default keeps data in RAM unless you use long trace mode to stream the data to disk.

In summary, you probably need to add:
write_into_file: true
file_write_period_ms: 5000

It's very very strongly recommended that you use RING_BUFFER configs for write_into_file traces: using DISCARD buffer is not well defined.

**Comment by LalitMaganti** [comment]
```
write_into_file: true
file_write_period_ms: 5000
```

Is just added at the top level of your trace config.

**Comment by forrestxu123** [comment]
write_into_file: true
file_write_period_ms: 5000

buffers: {
    size_kb: 1522240
    fill_policy: DISCARD
}
buffers: {
    size_kb: 4096
    fill_policy: DISCARD
}
data_sources: {
    config {
        name: "android.log"
        android_log_config {
        }
    }
}

data_sources: {
    config {
        name: "android.surfaceflinger.frametimeline"
    }
}

data_sources: {
    config {
        name: "android.packages_list"
        target_buffer: 1
    }
}
data_sources: {
    config {
        name: "linux.process_stats"
        target_buffer: 1
        process_stats_config {
            scan_all_processes_on_start: true
        }
    }
}
data_sources: {
    config {
        name: "linux.perf"
        perf_event_config {
            timebase {
                frequency: 4000
            }
            callstack_sampling {
                write_into_file: true
                file_write_period_ms: 5000
            }
        }
    }
}
data_sources: {
    config {
        name: "linux.sys_stats"
        sys_stats_config {
            stat_period_ms: 1000
            stat_counters: STAT_CPU_TIMES
            stat_counters: STAT_FORK_COUNT
        }
    }
}

data_sources: {
    config {
        name: "linux.ftrace"
        ftrace_config {
            ftrace_events: "sched/sched_switch"
            ftrace_events: "power/suspend_resume"
            ftrace_events: "sched/sched_wakeup"
            ftrace_events: "sched/sched_wakeup_new"
            ftrace_events: "sched/sched_waking"
            ftrace_events: "power/cpu_frequency"
            ftrace_events: "power/cpu_idle"
            ftrace_events: "raw_syscalls/sys_enter"
            ftrace_events: "raw_syscalls/sys_exit"
            ftrace_events: "sched/sched_process_exit"
            ftrace_events: "sched/sched_process_free"
            ftrace_events: "task/task_newtask"
            ftrace_events: "task/task_rename"
        }
    }
}
duration_ms: 40000
RIGHT?

**Comment by LalitMaganti** [comment]
Yes.

**Comment by forrestxu123** [comment]
-:49:17 error: No field named "write_into_file" in proto CallstackSampling
                write_into_file: true
                ^~~~~~~~~~~~~~~ 
-:50:17 error: No field named "file_write_period_ms" in proto CallstackSampling
                file_write_period_ms: 5000
                ^~~~~~~~~~~~~~~~~~~~

**Comment by forrestxu123** [comment]
is it only support on android14? i am working on android 12?

**Comment by LalitMaganti** [comment]
Not sure why you added to both the top level and in `callstack_sampling`: it's only valid at the top level. Please remove the field from the `callstack_sampling` field.

**Comment by forrestxu123** [comment]
sorry! mistake. working now. if i use RING_BUFFER , i cannot see callstack sample.  size_kb: 1522240 work for may emulator
